pkg/agent/controller/networkpolicy/allocator.go

```diff
@@ -63,24 +79,82 @@ func newIDAllocator(allocatedIDs ...uint32) *idAllocator {
 	return allocator
 }
 
-// allocate allocates an uint32 ID if there's available, otherwise error is returned.
-// It will try to reuse IDs that have been released first, then allocate a new ID by
-// incrementing the last allocated one.
-func (a *idAllocator) allocate() (uint32, error) {
+// allocateForRule allocates an uint32 ID for a given rule if it's available, otherwise
+// an error is returned. It will try to reuse the IDs that have been released first,
+// then allocate a new ID by incrementing the last allocated one.
+func (a *idAllocator) allocateForRule(rule *types.PolicyRule) error {
 	a.Lock()
 	defer a.Unlock()
 
 	if len(a.availableSlice) > 0 {
 		var id uint32
 		id, a.availableSlice = a.availableSlice[0], a.availableSlice[1:]
 		delete(a.availableSet, id)
-		return id, nil
+
+		// Add ID to the rule and the rule to asyncRuleCache.
+		rule.FlowID = id
+		a.asyncRuleCache.rules.Add(rule)
+
+		return nil
 	}
 	if a.lastAllocatedID == math.MaxUint32 {
-		return 0, fmt.Errorf("no ID available")
+		return fmt.Errorf("no ID available")
 	}
 	a.lastAllocatedID++
-	return a.lastAllocatedID, nil
+
+	// Add ID to the rule and the rule to asyncRuleCache.
+	rule.FlowID = a.lastAllocatedID
+	a.asyncRuleCache.rules.Add(rule)
+
+	return nil
+}
+
+// forgetRule adds the rule to the async delete queue with a given delay.
+func (a *idAllocator) forgetRule(ruleID uint32) {
+	a.asyncRuleCache.deleteQueue.AddAfter(ruleID, asyncDeleteInterval)
+}
+
+func (a *idAllocator) getRuleFromAsyncCache(ruleID uint32) (*types.PolicyRule, error) {
```

Comment:
maybe we should also return `(*types.PolicyRule, bool, error)` to match `GetByKey`?

Reply:
Good suggestion. Can distinguish error and a case with no rule in the cache.